stan/math/prim/mat/fun/log_mix.hpp

```diff
@@ -0,0 +1,99 @@
+
+#include <stan/math/prim/mat/meta/is_vector.hpp>
+#include <stan/math/prim/mat/meta/get.hpp>
+#include <stan/math/prim/mat/meta/length.hpp>
+#include <stan/math/prim/mat/fun/log_sum_exp.hpp>
+#include <stan/math/prim/mat/fun/log.hpp>
+#include <stan/math/prim/arr/meta/get.hpp>
+#include <stan/math/prim/arr/meta/length.hpp>
+#include <stan/math/prim/scal/err/check_bounded.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/meta/partials_return_type.hpp>
+#include <stan/math/prim/scal/meta/operands_and_partials.hpp>
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
+#include <stan/math/prim/scal/fun/value_of.hpp>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Return the log mixture density with specified mixing proportion
+     * and log densities.
+     *
+     * \f[
+     * \frac{\partial }{\partial p_x}
+     * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+     * \exp^{\log\left(p_n\right)+d_n}\right)
+     * =\frac{e^{d_x}}{e^{d_1}p_1+\cdot\cdot\cdot+e^{d_m}p_m}
+     * \f]
+     *
+     * \f[
+     * \frac{\partial }{\partial d_x}
+     * \log\left(\exp^{\log\left(p_1\right)+d_1}+\cdot\cdot\cdot+
+     * \exp^{\log\left(p_n\right)+d_n}\right)
+     * =\frac{e^{d_x}p_x}{e^{d_1}p_1+\cdot\cdot\cdot+e^{d_m}p_m}
+     * \f]
+     *
+     * @param theta vector of mixing proportions in [0, 1].
+     * @param lambda vector of log densities.
+     * @return log mixture of densities in specified proportion
+     */
+    template <typename T_theta, typename T_lam>
+    typename return_type<T_theta, T_lam>::type
+    log_mix(const T_theta& theta, const T_lam& lambda) {
+      static const char* function = "log_mix";
+      typedef typename stan::partials_return_type<T_theta, T_lam>::type
+        T_partials_return;
+
+      const int N = length(theta);
+
+      check_bounded(function, "theta", theta, 0, 1);
+      check_not_nan(function, "lambda", lambda);
+      check_not_nan(function, "theta", theta);
+      check_finite(function, "lambda", lambda);
+      check_finite(function, "theta", theta);
+      check_consistent_sizes(function, "theta", theta, "lambda", lambda);
+
+      Eigen::Matrix<T_partials_return, Eigen::Dynamic, 1> theta_dbl(N, 1);
+      scalar_seq_view<T_theta> theta_vec(theta);
+      for (int n = 0; n < N; ++n)
+        theta_dbl[n] = value_of(theta_vec[n]);
+
+      Eigen::Matrix<T_partials_return, Eigen::Dynamic, 1> lam_dbl(N, 1);
+      scalar_seq_view<T_lam> lam_vec(lambda);
+      for (int n = 0; n < N; ++n)
+        lam_dbl[n] = value_of(lam_vec[n]);
+
+      Eigen::Matrix<T_partials_return, Eigen::Dynamic, 1> logp_tmp(N, 1);
+      logp_tmp = log(theta_dbl) + lam_dbl;
+
+      T_partials_return logp = log_sum_exp(logp_tmp);
```

Comment:
I don't see another usage of `logp_tmp`, so I think this can all be replaced with a one-liner:
```
T_partials_return logp = log_sum_exp(log(theta_dbl) + lam_dbl);
```

The suffix `_tmp` should be avoided. Here, the variable could just be called `logp`.  All variables are "temp" in some sense and shorter is better all else being equal.

Reply:
I had to split this into two steps because the return type of ```log(theta_dbl) + lam_dbl``` is a 'CwiseBinaryOp', so log_sum_exp can't deal with it until Eigen assigns the result to a vector.

I'm not sure if there's a way of using the matrix/array wrappers to avoid this, but I haven't had any luck with figuring it out.